man/hbeatd.7
.ig
Copyright (c) 2012, Comfirm AB
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
 
    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.
    * Neither the name of the Comfirm AB nor the names of its contributors
      may be used to endorse or promote products derived from this software
      without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR NY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
..
.
.de TQ
.  br
.  ns
.  TP \\$1
..
.
.
.TH HBEATD 7 "13 December 2012" "Hbeatd Version 2.0.0"
.
.
.\" -----------------------------------------------------------------
.
.SH NAME
.
.B hbeatd
\-\- a simple and lightweight heartbeat sensor and pulse program
.
.
.\" -----------------------------------------------------------------
.
.SH SYNOPSIS
.
.B hbeatd
[
.IR options .\|.\|.\&
]
.br
.B hbeatd
[
.B -pvq\&
]
[
.B -d\&
.I ip
]
[
.B -P\&
.I port
]
[
.B -i\&
.I interval
]
[
.B -g\&
.I group
]
.br
.B hbeatd
[
.B -sv\&
]
[
.B -P\&
.I port
]
[
.B -t\&
.I tolerance
]
.
.
.\" -----------------------------------------------------------------
.
.SH DESCRIPTION
.
.
.B hbeatd
is a simple and lightweight program for sending and receiving
heartbeats that can be used to check the status of nodes in a cluster
of servers. The pulse program on the nodes sends udp packages to
chosen sensor. The sensor collects the ip's along with the server
group and builds a list. Everytime a new heartbeat is received the
list will be updated with a fresh timestamp and checked for dead
nodes (nodes with a timestamp older than the tolerance level). If a
node is dead, a script located at /etc/hbeatd/notify will run.
.
.\" -----------------------------------------------------------------
.
.SH OPTIONS
.
Unless specifically stated otherwise, options are applicable in all operating modes.
.TP
.B \-p
Run
.B hbeatd
in pulse mode.
.
.TP
.B \-s
Run
.B hbeatd
in sensor mode.
.
.TP
.B \-v
Run in verbose mode, not as daemon.
.
.TP
.BI \-q
Enable the http server to access status from a web API.
.
.TP
.BI \-d ip
Destination ip address (only available in pulse mode).
The default is 127.0.0.1.
.
.TP
.BI \-P port
Destination port number (only available in pulse mode).
The default is 6220.
.
.TP
.BI \-i interval
The interval in milliseconds between pulses (only available in pulse mode).
The default is 1000 (1 second).
.
.TP
.BI \-g group
A group name ex. 'web' or 'db', good when building lists of available nodes for deployment and maintenance etc. (only available in pulse mode).
The default is 'misc'.
.
.TP
.BI \-t tolerance
The number of seconds before reporting that the node is dead (only available in sensor mode).
.
The default is 4.
.
.
.SH "EXAMPLES"
.
.TP
hbeat -d 192.168.0.50
Start
.B hbeatd
in pulse mode.
.
.TP
hbeatd -p -i 800 -v
Start
.B hbeatd
in pulse mode without daemonizing it. Sending heartbeats with 800 milliseconds interval.
.
.TP
hbeatd -s
Start
.B hbeatd
in sensor mode.
.
.TP
hbeatd -s -t 20
Start
.B hbeatd
in sensor mode with a tolereance of 20 seconds.
.
.TP
hbeatd -s -P 8910 -v
Start
.B hbeatd
in sensor mode on port 8910 without daemonizing it (good for
debugging).
.
.\" -----------------------------------------------------------------
.
.SH FILES
.TP
.B /etc/hbeatd/notify
An executable file that runs everytime
.B hbeatd
founds a new, dead or resurrected node.
.
The script or program can take three arguments; status, ip and group.
.
.br
.B Example sh-script:

if [ "$1" = "dead" ]; then
.br
	echo "dead: $2($3)" >> /tmp/hbeatd.log
.br
elif [ "$1" = "up" ]; then
.br
	echo "up: $2($3)" >> /tmp/hbeatd.log
.br
elif [ "$1" = "new" ]; then
.br
	echo "new: $2($3)" >> /tmp/hbeatd.log
.br
fi
.
.TP
.B /etc/hbeatd/hbeatd.conf
Configuration file used when starting
.B hbeatd
from an init script.
.
.TP
.B /etc/init.d/hbeatd, /etc/rc.d/hbeatd
Init script for automatic start of
.B hbeatd
at boot. Can also be used for a more simple way to start
.B hbeatd
with the configurations in hbeatd.conf instead of passing them along
as arguments.
.
.\" -----------------------------------------------------------------
.
.SH AUTHOR
.
Jack Engqvist Johansson <jack.johansson@comfirm.se>
.
.\" Local Variables:
.\" mode: nroff
.\" End:


hbeatd.c
#include <arpa/inet.h>
#include <netinet/in.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <sys/mman.h>
#include <fcntl.h>
#include <unistd.h>
#include <semaphore.h>
#include <string.h>
#include <time.h>
#include <zlib.h>


#define HBEATD_VERSION "2.0.0"

#define NODE_COUNT 256
#define INT_SLEEP 1000
#define BUFLEN 256
#define SRV_GROUP "srv"
#define SRV_PORT 6220
#define SRV_IP "127.0.0.1"
#define TOLERANCE 4
#define SCRIPT_PATH "/etc/hbeatd/notify"


/* settings */
int pflag, sflag, vflag;
char *dvalue, *gvalue;
int Pvalue, rvalue, tvalue;

typedef struct {
	unsigned long int ip;
	unsigned long int time;
	unsigned long int uptime;
	unsigned long int beat;
	unsigned int live;
	char *groupname;
} node;

typedef struct {
	sem_t resource, mutex;
	node nodes[NODE_COUNT];
	unsigned int count;
	unsigned int heartbeats;
} container;

/* globals */
container *memory;
int fd, script_exists;

static void node_checker(void);
static void http_srv(void);
static void pulse(void);

static void die(char *s)
{
	fprintf(stderr, "error: %s\n", s);
	exit(1);
}
static void usage(void)
{
	(void)fprintf(stderr, "usage:\n");
	(void)fprintf(stderr, "\thbeatd [ -pv ] [ -d ip ] [ -P port ] [ -i interval ] [ -g group ]\n");
	(void)fprintf(stderr, "\thbeatd [ -sv ] [ -P port ] [ -t tolerance ]\n");

	exit(1);
}
static int fexists(char *fname)
{
	FILE *file;
	if ((file = fopen(fname, "r")) == NULL) {
		fprintf(stderr, "file %s could not be opened\n", fname);
	} else {
		fclose(file);
		return 1;
	}
	return 0;
}
void signal_handler(sig)
int sig;
{
	switch(sig) {
	case SIGHUP:
		printf("hangup signal catched");
		break;
	case SIGTERM:
		printf("terminate signal catched");
		exit(0);
		break;
	}
}
int msleep(unsigned long millisec)
{
    struct timespec req = { 0 };
    time_t sec = (int)(millisec / 1000);
    millisec = millisec - (sec * 1000);
    req.tv_sec = sec;
    req.tv_nsec = millisec * 1000000L;
    while(nanosleep(&req, &req) == -1)
         continue;

    return 1;
}

int64_t timespecDiff(struct timespec *timeA_p, struct timespec *timeB_p)
{
	  return ((timeA_p->tv_sec * 1000000000) + timeA_p->tv_nsec) -
		             ((timeB_p->tv_sec * 1000000000) + timeB_p->tv_nsec);
}



int main(int argc, char *argv[])
{
	(void)printf("hbeatd version %s\n%s\n\n", HBEATD_VERSION, "Copyright (c) 2012 Comfirm AB");

	int c;
	pflag = sflag = vflag = 0;
	dvalue = NULL;
	gvalue = NULL;
	Pvalue = rvalue = tvalue = 0;
	script_exists = 0;
	unsigned int iserver = 0;

	while ((c = getopt(argc, argv, "spvhqd:P:i:g:t:")) != -1)
	{
		switch (c) {
			case 's':
				sflag = 1;
				pflag = 0;
				break;
			case 'p':
				pflag = 1;
				sflag = 0;
				break;
			case 'v':
				vflag = 1;
				break;
			case 'h':
				usage();
				break;
			case 'd':
			     dvalue = optarg;
			     break;
			case 'P':
			     Pvalue = atoi(optarg);
			     break;
		     	case 'i':
			     rvalue = atoi(optarg);
			     break;
			case 'g':
			     gvalue = optarg;
			     break;
			case 't':
			     tvalue = atoi(optarg);
			     break;
			case 'q':
				 iserver = 1;
			     break;
			case '?':
			default:
				usage();
		}
	}

	if(dvalue == NULL)
		dvalue = SRV_IP;
	if(Pvalue == 0)
		Pvalue = SRV_PORT;
	if(rvalue == 0)
		rvalue = INT_SLEEP;
	if(tvalue == 0)
		tvalue = TOLERANCE;
	if(gvalue == NULL)
		gvalue = SRV_GROUP;

	if(!sflag)
	{
		printf("PULSE MODE\n");
		printf("sending heartbeats to %s:%d(%s)...\n", dvalue, Pvalue, gvalue);
	}
	else
	{
		printf("SENSOR MODE\n");
		printf("listening on *:%d...\n", Pvalue);
	}

	/* daemonize */
	if(!vflag)
	{
		int f;

		/* close all descriptors */
		for (f = getdtablesize(); f >= 0; --f)
			close(f);

		/* fork */
		f = fork();
		if (f < 0)
			die("could not fork process");

		/* exit parent */
		if (f > 0)
			exit(0);

		/* continue */
		setsid();
		umask(027);
		chdir("/");

		/* handle signals */
		signal(SIGCHLD, SIG_IGN); /* ignore child */
		signal(SIGTSTP, SIG_IGN); /* ignore tty signals */
		signal(SIGTTOU, SIG_IGN);
		signal(SIGTTIN, SIG_IGN);
		signal(SIGHUP, signal_handler); /* catch hangup signal */
		signal(SIGTERM, signal_handler); /* catch kill signal */

		/* done */
	}

	/* memory sharing */
	fd = shm_open("/tmp/hbeatd-shm.tmp", O_RDWR|O_CREAT, 0777);

	if (fd == -1) {
		die("shm_open failed");
	}

	ftruncate(fd, sizeof(container));
	memory = mmap(NULL, sizeof(container), PROT_READ|PROT_WRITE, MAP_SHARED, fd, 0);

	/* http info server? */
	if(iserver == 1 && sflag == 1)
		http_srv();

	/* pulse mode */
	if(!sflag)
	{
		(void)pulse();
		exit(0);
	}

	/* sensor mode */
	if(fexists(SCRIPT_PATH))
	{
		script_exists = 1;
	}

	/* socket */
	struct sockaddr_in sock, si_other;
	struct in_addr addr;

	int s, slen = sizeof(si_other);
	char buf[BUFLEN];
	int buf_len = 0;

	if((s = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) == -1)
		die("couldn't create socket");

	memset((char *) &sock, 0, sizeof(sock));
	memset((char *) &addr, 0, sizeof(addr));

	sock.sin_family = AF_INET;
	sock.sin_port = htons(Pvalue);
	sock.sin_addr.s_addr = htonl(INADDR_ANY);

	if(bind(s, (struct sockaddr *)&sock, sizeof(sock)) == -1)
		die("bind() failed");

	/* fork and check nodes */
	node_checker();


	time_t seconds;
	unsigned int i, n, count;
	unsigned long int time_now;
	int add = 1;
	node *nodes = memory->nodes;

	count = 0;
	memory->count = 0;
	memory->heartbeats = 0;

	while(1)
	{
		for(n = 0; n <= buf_len; n++) {
			buf[n] = (char)0;
		}

		if(recvfrom(s, buf, BUFLEN, 0, (struct sockaddr *)&si_other, &slen) == -1)
			die("recvfrom() failed");

		memory->heartbeats = memory->heartbeats + 1;

		if(vflag)
			printf("recieved heartbeat from %s\n", inet_ntoa(si_other.sin_addr));

		buf_len = strlen(buf);

		/* get current time */
		time_now = (unsigned long int)time(NULL);

		/* first, search for it in the list */
		add = 1;
		for(i = 0; i < count; i++)
		{
			/* is it there? */
			if(nodes[i].ip == si_other.sin_addr.s_addr)
			{
				/* resurrected from the dead? :0 */
				if(nodes[i].live == 0)
				{
					/* update groupname */
					char *groupname = malloc(buf_len + 1);
					memcpy(groupname, buf, buf_len);

					if(nodes[i].groupname != NULL)
						free(nodes[i].groupname);
					nodes[i].groupname = groupname;

					/* update the uptime */
					nodes[i].uptime = time_now;

					/* spread the good news... */
					pid_t pID = fork();
					if (pID == 0)
					{
						addr.s_addr = nodes[i].ip;
						char *ip_str = inet_ntoa(addr);

						if(vflag)
							printf("resurrected: %s(%s)\n", ip_str, nodes[i].groupname);

						if (script_exists)
							execl(SCRIPT_PATH, SCRIPT_PATH, "up", ip_str, nodes[i].groupname, (char *)0);
						exit(0);
					}
					nodes[i].live = 1;
				}

				/* update last seen time*/
				nodes[i].time = time_now;
				nodes[i].beat = 1;
				add = 0;
			}
		}

		if(add)
		{
			/* set group name */
			char *groupname = malloc(buf_len + 1);
			memcpy(groupname, buf, buf_len);

			node node_new = { si_other.sin_addr.s_addr, time_now, time_now, 1, 1, groupname };
			nodes[count] = node_new;

			/* set the uptime */
			nodes[i].uptime = time_now;

			count = count + 1;
			memory->count = count;

			/* do something (run script) */
			pid_t pID = fork();
			if (pID == 0)
			{
				addr.s_addr = nodes[i].ip;
				char *ip_str = inet_ntoa(addr);

				if(vflag)
					printf("new: %s(%s)\n", ip_str, node_new.groupname);

				if (script_exists)
					execl(SCRIPT_PATH, SCRIPT_PATH, "new", ip_str, nodes[i].groupname, (char *)0);
				exit(0);
			}
		}
	}

	close(s);
	return 0;
}

static void node_checker(void)
{
	/* fork so the node checker can work alone */
	pid_t pID = fork();
	if (pID == 0)
	{
		int i, count;
		unsigned long int time_now;
		struct in_addr addr;
		node *nodes = memory->nodes;

		while(1)
		{
			time_now = (unsigned long int)time(NULL);
			count = memory->count;

			for(i = 0; i < count; i++)
			{
				if(nodes[i].live == 1 && time_now - nodes[i].time >= tvalue)
				{
					nodes[i].live = 0;

					/* do something (run script) */
					pid_t pID = fork();
					if (pID == 0)
					{
						addr.s_addr = nodes[i].ip;
						char *ip_str = inet_ntoa(addr);

						if(vflag)
							printf("dead: %s(%s)\n", ip_str, nodes[i].groupname);

						if (script_exists)
							execl(SCRIPT_PATH, SCRIPT_PATH, "dead", ip_str, nodes[i].groupname, (char *)0);
						exit(0);
					}
				}
			}

			msleep(1000);
		}

		exit(0);
	}
}

static void http_srv(void)
{
	/* fork */
	pid_t pid = fork();
	if (pid == 0)
	{
		char mesg[99999], *reqline[3], data_to_send[1024], path[99999];
		int rcvd, fd, bytes_read;

		char *response;
		response = malloc(sizeof(int) * 2200);

		char *body;
		body = malloc(sizeof(int) * 2000);

		char *headers;
		headers = malloc(sizeof(int) * 200);

		int q, w, index, ip_len, len_headers, yes;
		char buffer[50];
		unsigned long int time_now;

		/* listen on traffic */
		struct sockaddr_storage their_addr;
		socklen_t addr_size;
		char s[INET6_ADDRSTRLEN];

		struct addrinfo hints, *serverinfo, *p;
		int sockfd, sockfd_new;

		memset(&hints, 0, sizeof hints);
		hints.ai_family = AF_INET;
		hints.ai_socktype = SOCK_STREAM;
		hints.ai_flags = AI_PASSIVE;

		if(getaddrinfo(NULL, "3490", &hints, &serverinfo) != 0) {
			die("getaddrinfo() failed");
		}

		for(p = serverinfo; p != NULL; p = p->ai_next) {
			if((sockfd = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) == -1) {
				fprintf(stderr, "error: socket() failed\n");
				continue;
			}

			yes = 1;
			if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(int)) == -1) {
				fprintf(stderr, "error: setsockopt() failed\n");
				continue;
			}

			if(bind(sockfd, p->ai_addr, p->ai_addrlen) == -1) {
				close(sockfd);
				fprintf(stderr, "error: bind() failed\n");
				continue;
			}

			break;
		}

		if(p == NULL) {
			die("bind() failed");
		}

		freeaddrinfo(serverinfo);

		if(listen(sockfd, 5) == -1) {
			die("listen() failed");
		}

		// reap all dead processes??

		printf("info server: started\n");

		while(1) {
			/* accept the connection */
			addr_size = sizeof(their_addr);
			sockfd_new = accept(sockfd, (struct sockaddr *)&their_addr, &addr_size);

			/* time measurement */
			struct timespec start, end;
			clock_gettime(CLOCK_MONOTONIC, &start);


			if(sockfd_new == -1) {
				fprintf(stderr, "accept() failed\n");
				continue;
			}

			memset((void*)mesg, (int)'\0', 99999);
			rcvd = recv(sockfd_new, mesg, 99999, 0);

			/* get current time */
			time_now = (unsigned long int)time(NULL);


			/* check http request */
			if(strlen(mesg) < 3 ) {
				/* bad request */
				sprintf(headers, "%s", "HTTP/1.1 400 Bad Request\r\nServer: hbeatd/2.0.0\r\nAge: 0\r\nAccess-Control-Allow-Origin: *\r\nAllow: GET\r\nContent-length: 0");

				if(send(sockfd_new, headers, strlen(headers), 0) == -1) {
					fprintf(stderr, "send() failed\n");
					close(sockfd_new);
				}
				close(sockfd_new);
				continue;
			}
			else if(strncmp("OPTIONS ", mesg, 8) == 0) {
				/* OPTIONS */
				sprintf(headers, "%s", "HTTP/1.1 200 OK\r\nServer: hbeatd/2.0.0\r\nAge: 0\r\nAccess-Control-Allow-Origin: *\r\nAllow: OPTIONS,GET\r\nContent-length: 0\r\n\r\n");

				if(send(sockfd_new, headers, strlen(headers), 0) == -1) {
					fprintf(stderr, "send() failed\n");
					close(sockfd_new);
				}
				close(sockfd_new);
				continue;
			}
			else if(strncmp("GET ", mesg, 4) != 0) {
				/* invalid method */
				sprintf(headers, "%s", "HTTP/1.1 405 Method Not Allowed\r\nServer: hbeatd/2.0.0\r\nAge: 0\r\nAccess-Control-Allow-Origin: *\r\nAllow: GET\r\nContent-length: 0\r\n\r\n");

				if(send(sockfd_new, headers, strlen(headers), 0) == -1) {
					fprintf(stderr, "send() failed\n");
					close(sockfd_new);
				}
				close(sockfd_new);
				continue;
			}

			/* GET */

			index = 0;

			/* headers */
			sprintf(headers, "%s", "HTTP/1.1 200 OK\r\nServer: hbeatd/2.0.0\r\nCache-Control: no-cache\r\nAge: 0\r\nAccess-Control-Allow-Origin: *\r\nContent-type: application/json\r\nContent-length: 000");

			body[index++] = '\r';
			body[index++] = '\n';
			body[index++] = '\r';
			body[index++] = '\n';

			/* start json */
			body[index++] = '{';

			/* uptime seen */
			body[index++] = '\"';
			body[index++] = 'h';
			body[index++] = '\"';
			body[index++] = ':';
			sprintf(buffer, "%d", memory->heartbeats);
			for(w = 0; w < strlen(buffer); w++) {
				body[index++] = buffer[w];
			}
			body[index++] = ',';

			body[index++] = '\"';
			body[index++] = 'n';
			body[index++] = '\"';
			body[index++] = ':';
			body[index++] = '[';

			/* node info json */
			for(q = 0; q < memory->count; q++) {
				struct in_addr addr2;
				addr2.s_addr = memory->nodes[q].ip;
				char *ip_str = inet_ntoa(addr2);

				if(q != 0)
					body[index++] = ',';

				body[index++] = '{';

				/* ip */
				body[index++] = '\"';
				body[index++] = 'i';
				body[index++] = '\"';
				body[index++] = ':';
				body[index++] = '\"';
				for(w = 0; w < strlen(ip_str); w++) {
					body[index++] = ip_str[w];
				}
				body[index++] = '\"';
				body[index++] = ',';

				/* last seen */
				body[index++] = '\"';
				body[index++] = 't';
				body[index++] = '\"';
				body[index++] = ':';
				sprintf(buffer, "%d", time_now - memory->nodes[q].time);
				for(w = 0; w < strlen(buffer); w++) {
					body[index++] = buffer[w];
				}
				body[index++] = ',';

				/* uptime */
				body[index++] = '\"';
				body[index++] = 'u';
				body[index++] = '\"';
				body[index++] = ':';
				sprintf(buffer, "%d", memory->nodes[q].time - memory->nodes[q].uptime);
				for(w = 0; w < strlen(buffer); w++) {
					body[index++] = buffer[w];
				}
				body[index++] = ',';

				/* heartbeat */
				body[index++] = '\"';
				body[index++] = 'b';
				body[index++] = '\"';
				body[index++] = ':';
				if(memory->nodes[q].beat > 0) {
					body[index++] = '1';
					memory->nodes[q].beat = 0;
				}
				else {
					body[index++] = '0';
				}
				body[index++] = ',';

				/* live */
				body[index++] = '\"';
				body[index++] = 'l';
				body[index++] = '\"';
				body[index++] = ':';
				if(memory->nodes[q].live == 1) {
					body[index++] = '1';
				}
				else {
					body[index++] = '0';
				}

				body[index++] = '}';
			}

			/* end json */
			body[index++] = ']';
			body[index++] = '}';
			body[index++] = (char)0;

			/* put it together */
			sprintf(response, "%s%d%s", headers, strlen(body) - 4, body);

			//msleep(200);

			if(send(sockfd_new, response, strlen(response), 0) == -1) {
				fprintf(stderr, "send() failed\n");
				close(sockfd_new);
			}
			close(sockfd_new);

			/* time measurements */
			clock_gettime(CLOCK_MONOTONIC, &end);
			uint64_t timeElapsed = timespecDiff(&end, &start);
			printf("[http_src] request handled in: %f milliseconds\n", (double)((double)timeElapsed / 1000000.00));

		}
		close(sockfd);

		/* clean up */
		free(response);
		free(body);
		free(headers);

		die("information server started.... nothing to do, quiting!\n");
		exit(0);
	}
	else {
		return;
	}
}

static void pulse(void)
{
	struct sockaddr_in sock;
	int s, i, slen = sizeof(sock);

	if ((s = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) == -1)
		die("couldn't create socket");

	memset((char *) &sock, 0, sizeof(sock));
	sock.sin_family = AF_INET;
	sock.sin_port = htons(Pvalue);

	if (inet_aton(dvalue, &sock.sin_addr) == 0)
		die("inet_aton() failed");

	while (1)
	{
		if(vflag)
			printf("[ heartbeat ]\n");

		if (sendto(s, gvalue, strlen(gvalue), 0, (struct sockaddr *)&sock, slen) == -1)
			die("failed to send");


		msleep(rvalue);
	}

	close(s);
}

